pkg/storer/internal/chunkstore/chunkstore.go

```diff
@@ -33,6 +33,175 @@ const RetrievalIndexItemSize = swarm.HashSize + 8 + sharky.LocationSize + 4
+	err := s.Read(ctx, rIdx.Location, buf)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"chunk store: failed reading location: %v for chunk %s from sharky: %w",
+			rIdx.Location, rIdx.Address, err,
+		)
+	}
+
+	return swarm.NewChunk(rIdx.Address, buf), nil
+}
+
+func Has(_ context.Context, r storage.Reader, addr swarm.Address) (bool, error) {
+	return r.Has(&RetrievalIndexItem{Address: addr})
+}
+
+func Put(ctx context.Context, s storage.IndexStore, sh storage.Sharky, ch swarm.Chunk) error {
+	var (
+		rIdx = &RetrievalIndexItem{Address: ch.Address()}
+		loc  sharky.Location
+	)
+	err := s.Get(rIdx)
+	switch {
+	case errors.Is(err, storage.ErrNotFound):
+		// if this is the first instance of this address, we should store the chunk
+		// in sharky and create the new indexes.
+		loc, err = sh.Write(ctx, ch.Data())
+		if err != nil {
+			return fmt.Errorf("chunk store: write to sharky failed: %w", err)
+		}
+		rIdx.Location = loc
+		rIdx.Timestamp = uint64(time.Now().Unix())
+	case err != nil:
+		return fmt.Errorf("chunk store: failed to read: %w", err)
+	}
+
+	rIdx.RefCnt++
+
+	return s.Put(rIdx)
+}
+
+func Delete(ctx context.Context, s storage.IndexStore, sh storage.Sharky, addr swarm.Address) error {
+	rIdx := &RetrievalIndexItem{Address: addr}
+	err := s.Get(rIdx)
+	switch {
+	case errors.Is(err, storage.ErrNotFound):
+		return nil
+	case err != nil:
+		return fmt.Errorf("chunk store: failed to read retrievalIndex for address %s: %w", addr, err)
+	default:
+		rIdx.RefCnt--
+	}
+
+	if rIdx.RefCnt > 0 { // If there are more references for this we don't delete it from sharky.
+		err = s.Put(rIdx)
+		if err != nil {
+			return fmt.Errorf("chunk store: failed updating retrievalIndex for address %s: %w", addr, err)
+		}
+		return nil
+	}
+
+	return errors.Join(
+		sh.Release(ctx, rIdx.Location),
+		s.Delete(rIdx),
+	)
+}
+
+func Iterate(ctx context.Context, s storage.IndexStore, sh storage.Sharky, fn storage.IterateChunkFn) error {
+	return s.Iterate(
+		storage.Query{
+			Factory: func() storage.Item { return new(RetrievalIndexItem) },
+		},
+		func(r storage.Result) (bool, error) {
+			ch, err := readChunk(ctx, sh, r.Entry.(*RetrievalIndexItem))
+			if err != nil {
+				return true, err
+			}
+			return fn(ch)
+		},
+	)
+}
+
+func IterateChunkEntries(st storage.Reader, fn func(swarm.Address, uint32) (bool, error)) error {
+	return st.Iterate(
+		storage.Query{
+			Factory: func() storage.Item { return new(RetrievalIndexItem) },
+		},
+		func(r storage.Result) (bool, error) {
+			item := r.Entry.(*RetrievalIndexItem)
+			addr := item.Address
+			return fn(addr, item.RefCnt)
+		},
+	)
+}
+
+type LocationResult struct {
+	Err      error
+	Location sharky.Location
+}
+
+type IterateResult struct {
+	Err  error
+	Item *RetrievalIndexItem
+}
+
+// IterateLocations iterates over entire retrieval index and plucks only sharky location.
+func IterateLocations(
+	ctx context.Context,
+	st storage.Reader,
+	locationResultC chan<- LocationResult,
```

Comment:
As the goroutine that is created in here takes the full ownership of the locationResultC channel (it writes to it and closes the channel when it it done), it would be better to protect the channel as returning it as the read only channel. In all useasges in this pr, the channel is creaded before the IterateLocations() call and only read from it.

Reply:
done